File: languages/kotlin/binary-trees.kt
class Node(val left: Node?, val right: Node?)

fun make(depth: Int): Node {
    if (depth == 0) return Node(null, null)
    return Node(make(depth - 1), make(depth - 1))
}

fun check(node: Node): Long {
    if (node.left == null) return 1L
    return 1L + check(node.left) + check(node.right!!)
}

fun binaryTrees(n: Int): Long {
    val minDepth = 4
    val maxDepth = maxOf(minDepth + 2, n)
    val stretchDepth = maxDepth + 1

    var total = check(make(stretchDepth))
    val longLived = make(maxDepth)

    var depth = minDepth
    while (depth <= maxDepth) {
        val iterations = 1 shl (maxDepth - depth + minDepth)
        var s = 0L
        for (i in 0 until iterations) {
            s += check(make(depth))
        }
        total += s
        depth += 2
    }

    total += check(longLived)
    return total
}

fun main(args: Array<String>) {
    val n = if (args.isNotEmpty()) args[0].toInt() else 10
    println(binaryTrees(n))
    println("binary-trees($n)")
}

File: languages/ruby/binary-trees.rb
MIN_DEPTH = 4

# Real heap node: a plain Ruby object with two child references (idiomatic),
# no node pool / arena / flat-array encoding.
class Node
  attr_accessor :left, :right

  def initialize(left, right)
    @left = left
    @right = right
  end
end

def make(depth)
  return Node.new(nil, nil) if depth == 0
  Node.new(make(depth - 1), make(depth - 1))
end

def check(node)
  return 1 if node.left.nil?
  1 + check(node.left) + check(node.right)
end

def binary_trees(n)
  max_depth = MIN_DEPTH + 2 > n ? MIN_DEPTH + 2 : n
  stretch_depth = max_depth + 1

  total = check(make(stretch_depth))
  long_lived = make(max_depth)

  depth = MIN_DEPTH
  while depth <= max_depth
    iterations = 1 << (max_depth - depth + MIN_DEPTH)
    s = 0
    iterations.times { s += check(make(depth)) }
    total += s
    depth += 2
  end

  total += check(long_lived)
  total
end

n = ARGV[0] ? ARGV[0].to_i : 14
puts binary_trees(n)
puts "binary-trees(#{n})"
